theano/gpuarray/tests/test_dnn.py

```diff
@@ -2298,3 +2298,225 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+
+    # Transformation matrix
+    transform = [[-1, 0, 0],
+                 [0, -1, 0]]
+    theta = np.asarray(img_dims[0] * [transform], dtype=theano.config.floatX)
+
+    # Create symbolic variables for inputs and transformations
+    t_img = T.tensor4('img')
+    t_theta = T.tensor3('theta')
+
+    st_dnn = dnn.dnn_spatialtf(t_img, t_theta,
+                               scale_height=scale_height,
+                               scale_width=scale_width)
+    st_dnn_func = theano.function([t_img, t_theta], st_dnn)
+    # Check if function graph contains the spatial transformer's grid and sampler Ops
+    apply_nodes = st_dnn_func.maker.fgraph.apply_nodes
+    assert any([isinstance(node.op, dnn.GpuDnnTransformerGrid) for node in apply_nodes])
+    assert any([isinstance(node.op, dnn.GpuDnnTransformerSampler) for node in apply_nodes])
+
+    img_out_gpu = st_dnn_func(img, theta)
+    img_out_gpu = np.asarray(img_out_gpu)
+
+    # Setup CPU Op
+    st_cpu = spatialtf_cpu(t_img, t_theta, scale_height, scale_width, 'nearest')
+    st_cpu_func = theano.function([t_img, t_theta], st_cpu, mode=mode_without_gpu)
+    img_out_cpu = st_cpu_func(img, theta)
+
+    atol, rtol = None, None  # use default absolute and relative error tolerances
+    if theano.config.floatX == 'float16':  # float16 requires higher tolerances
+        atol, rtol = 1e-0, 1e-2
+    utt.assert_allclose(img_out_cpu, img_out_gpu, atol=atol, rtol=rtol)
+
+
+def test_dnn_spatialtf_grad():
+    if not dnn.dnn_available(test_ctx_name):
+        raise SkipTest(dnn.dnn_available.msg)
+
+    utt.seed_rng()
+
+    inputs = T.tensor4('inputs')
+    theta = T.tensor3('theta')
+
+    out = dnn.dnn_spatialtf(inputs, theta, scale_height=0.25, scale_width=0.75)
+    out_mean = T.mean(out)
+    mean_gi = T.grad(out_mean, [inputs])
+    mean_gt = T.grad(out_mean, [theta])
+
+    f_gi = theano.function([inputs, theta], mean_gi)
+    assert any([isinstance(node.op, dnn.GpuDnnTransformerGradI)
+                for node in f_gi.maker.fgraph.apply_nodes])
+
+    f_gt = theano.function([inputs, theta], mean_gt)
+    assert any([isinstance(node.op, dnn.GpuDnnTransformerGradT)
+                for node in f_gt.maker.fgraph.apply_nodes])
+
+    # Generate random set of RGB images(pixel values in [0, 255]). The set of
+    # images is generated in the expected (NCHW) format
+    input_dims = (5, 3, 16, 16)
+    inputs_val = np.random.randint(low=0, high=256, size=input_dims).astype(theano.config.floatX)
+    # Tensor with transformations
+    theta_val = np.random.random((input_dims[0], 2, 3)).astype(theano.config.floatX)
+
+    # Check that the gradients are computed
+    f_gi(inputs_val, theta_val)
+    f_gt(inputs_val, theta_val)
+
+    def grad_functor(inputs, theta):
+        out = dnn.dnn_spatialtf(inputs, theta)
+        return out
+
+    atol, rtol = 1e-3, 1e-3
+    if theano.config.floatX == 'float32':  # use higher error tolerances with float32
+        atol, rtol = 1e-0, 1e-1
```

Comment:
I've done that, and because it is using inputs in the range [0,1], the tolerance of 5e-2 is now sufficient, even on float16.